packages/runtime/container-runtime/src/summarizerHeuristics.ts

```diff
@@ -87,6 +119,42 @@ export class SummarizeHeuristicRunner implements ISummarizeHeuristicRunner {
     }
 
     public dispose() {
-        this.idleTimer.clear();
+        this.idleTimer?.clear();
     }
 }
+
+export class MaxTimeSummarizeHeuristicStrategy implements ISummarizeHeuristicStrategy {
+    public readonly summarizeReason: Readonly<SummarizeReason> = "maxTime";
+
+    public shouldRunSummarize(configuration: ISummaryConfiguration, heuristicData: ISummarizeHeuristicData): boolean {
+        const timeSinceLastSummary = Date.now() - heuristicData.lastSuccessfulSummary.summaryTime;
+        return timeSinceLastSummary > configuration.maxTime;
+    }
+}
+
+export class WeightedOpsSummarizeHeuristicStrategy implements ISummarizeHeuristicStrategy {
+    public readonly summarizeReason: Readonly<SummarizeReason> = "maxOps";
+
+    constructor(
+        private readonly weightConfiguration: ISummarizeHeuristicWeightConfiguration,
+    ) { }
+
+    public shouldRunSummarize(configuration: ISummaryConfiguration, heuristicData: ISummarizeHeuristicData): boolean {
+        const weightedNumOfOps = (this.weightConfiguration.systemOpWeight    * heuristicData.numSystemOps)
+                               + (this.weightConfiguration.nonSystemOpWeight * heuristicData.numNonSystemOps);
+        return weightedNumOfOps > configuration.maxOps;
+    }
+}
+
+const DefaultHeuristicWeightConfiguration: ISummarizeHeuristicWeightConfiguration = {
```

Comment:
I'd prefer not to have these properties be added to ISummaryConfiguration.
Please note that after #9965 (at some future POV in time), we should move ISummaryConfiguration out of protocol-definitions into container runtime. We can do it even now, duplicating interface and starting to use new definition in runtime. I've opened #9990 to partially track that and rationalization of interfaces, but we can do move part now.